Intersect.Client/Interface/Game/EventWindow.cs

```diff
@@ -317,6 +384,140 @@ public void EventResponse1_Clicked(Base sender, ClickedEventArgs arguments)
+        private List<Label> Labels { get; set; }
+        private string[] Lines { get; set; }
+        private Label CurrentLabel => Labels.ElementAtOrDefault(LineIdx);
+        private string CurrentLine => Lines.ElementAtOrDefault(LineIdx);
+        private int LineIdx { get; set; }
+        private int CharIdx { get; set; }
+        private char? LastChar { get; set; }
+        private long LastUpdateTime;
+        private long LastPauseTime { get; set; }
+
+        public bool Done { get; private set; }
+
+        public long DoneAt { get; set; }
+
+        public void CarriageReturn()
+        {
+            LineIdx++;
+            if (LineIdx >= Lines.Length)
+            {
+                End();
+                return;
+            }
+            CharIdx = 0;
+            LastChar = null;
+        }
+
+        public void Initialize(List<Label> labels)
+        {
+            LastUpdateTime = Timing.Global.Milliseconds;
+            Labels = labels;
+            Lines = Labels.Select(l => l.Text).ToArray();
+            Labels.ForEach(l => l.SetText("")); // Clear the text from the labels. The writer is the captain now
+            LineIdx = 0;
+            CharIdx = 0;
+            LastChar = null;
+            Done = false;
+        }
+
+        public void Write(string voice)
+        {
+            if (Done)
+            {
+                return;
+            }
+
+            if (string.IsNullOrEmpty(CurrentLine) || CurrentLabel == default)
+            {
+                End();
+                return;
+            }
+
+            if (Timing.Global.Milliseconds < LastUpdateTime)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(voice) && CharIdx % ClientConfiguration.Instance.TypewriterSoundFrequency == 0)
+            {
+                Audio.StopAllGameSoundsOf(ClientConfiguration.Instance.TypewriterSounds.ToArray());
+                Audio.AddGameSound(voice ?? string.Empty, false);
+            }
+
+            CharIdx++;
+            if (CharIdx >= CurrentLine.Length)
+            {
+                CurrentLabel?.SetText(CurrentLine);
+                CarriageReturn();
+                return;
+            }
+
+            LastChar = CurrentLine[CharIdx - 1];
+
+            var written = CurrentLine?.Substring(0, CharIdx) ?? string.Empty;
+            CurrentLabel?.SetText(written);
+
+            LastUpdateTime = Timing.Global.Milliseconds + GetTypingSpeed();
+        }
+
+        private long GetTypingSpeed()
+        {
+            if (LastChar == null)
+            {
+                return TypingSpeed;
+            }
+
+            var currentChar = CurrentLine.ElementAtOrDefault(CharIdx);
+            if ((LastChar == '.' || LastChar == '?' || LastChar == '!' || LastChar == ':') && currentChar != LastChar)
+            {
+                return FullStopSpeed;
+            }
+            if ((LastChar == '-' || LastChar == ',' || LastChar == '-') && currentChar != LastChar)
```

Comment:
You check for hyphen twice here, or is it two look-a-likes? Shouldn't semi-colon also be listed?